### PR TITLE
Rework management of Salt jobs in the UI

### PR DESCRIPTION
**Pull request by gdemonet**
**Component**: ui

**Context**:

As the UI will need to handle more and more long-running Salt operations (such as importing a Solution or preparing an Environment), the pattern introduced for deployment of Nodes needs to be extended.

**Summary**:

- extract all logic related to management of Salt jobs from `app.nodes` to `app.salt`
- refine the pattern to make it more generic: a deployment job is now named `deploy-node/<node-name>` to avoid future collisions
- store the completion time of a job in localStorage, to provide efficient reloading of the application
- keep jobs in localStorage after their completion and implement a garbage collection saga to clean them up after some time
- trigger notifications from the related duck (here, `app.nodes`), while the action gets sent from the `app.salt` duck
- rework the NodeDeployment page slightly (simplification and handling of some edge-cases)

(read commit messages for more info)

**Acceptance criteria**: 

Everything still works, and one can now find info about previously failed jobs even after a refresh. Unit tests were also slightly reworked to make use of features of `jest` to avoid code duplication - make sure you (UI devs) understand how this works and feel comfortable with it.

### CODE REVIEW

File: ui/src/ducks/app/salt.js [hunk 1]
```diff
@@ -1,31 +1,255 @@
...
+  const savedJobs = listJobsFromLocalStorage();
+  yield all(savedJobs.map(job => put(addJobAction(job))));
+  yield all(savedJobs.map(job => call(refreshJobStatus, job)));
+}
+
+function* manageLocalStorage() {
+  // NOTE: instead of creating a saga per action type, since the operations on
+  // localStorage rely on its previous state, we need to execute them serially.
+  // We don't set up a buffer size as operations should be infrequent and fast.
+  const channel = actionChannel([ADD_JOB, REMOVE_JOB, JOB_COMPLETED]);
```
Comment (MonPote) [inline ui/src/ducks/app/salt.js]: Very cool pattern :)

File: ui/src/containers/NodeDeployment.js [hunk 2]
```diff
@@ -140,14 +129,26 @@ const NodeDeployment = ({ intl }) => {
...
-          <NodeDeploymentStatus>
-            <Steppers steps={steps} activeStep={activeStep} />
-          </NodeDeploymentStatus>
-          <NodeDeploymentEvent>
-            <ReactJson src={events} name={nodeId} collapsed={true} />
-          </NodeDeploymentEvent>
-        </NodeDeploymentContent>
+        {activeJob === undefined ? (
+          <InfoMessage>
+            {`No deployment job found for Node ${nodeName}.`}
```
Comment (carlito-scality) [inline ui/src/containers/NodeDeployment.js]: translation here

File: ui/src/containers/NodeDeployment.js [hunk 3]
```diff
@@ -140,14 +129,26 @@ const NodeDeployment = ({ intl }) => {
...
+          </InfoMessage>
+        ) : activeJob.completed && isEmpty(activeJob.status) ? (
+          <InfoMessage>{`Refreshing job info...`}</InfoMessage>
```
Comment (carlito-scality) [inline ui/src/containers/NodeDeployment.js]: translation here

File: ui/src/ducks/app/nodes.js [hunk 4]
```diff
@@ -434,104 +361,60 @@ export function* deployNode({ payload }) {
...
-      jobs.map(jid => {
-        if (data.tag.includes(jid)) {
-          return call(updateDeployEvents, jid, data);
-        }
-        return data;
-      })
+    yield put(
+      addJobAction({
+        name: `deploy-node/${payload.name}`,
+        jid: result.return[0].jid,
```
Comment (carlito-scality) [inline ui/src/ducks/app/nodes.js:371]: why don't we add `node: payload.name` to avoid `job.name.replace(/^(deploy-node\/)/, '')`?

File: ui/src/services/salt/utils.js [hunk 5]
```diff
@@ -1,7 +1,50 @@
...
+  }
+}
+
+export function markJobCompleteInLocalStorage(jid, completedAt) {
+  const jobs = listJobsFromLocalStorage();
+  const job = jobs.find(item => item.jid === jid);
+  if (job) {
+    job.completedAt = completedAt;
+  }
+  localStorage.setItem(JOBS, JSON.stringify(jobs));
```
Comment (carlito-scality) [inline ui/src/services/salt/utils.js]: to be replaced by `saveJobsToLocalStorage(jobs)`

File: ui/src/ducks/app/salt.js [hunk 6]
```diff
@@ -1,31 +1,258 @@
...
+        break;
+    }
+  }
+}
+
+export function* refreshJobStatus(job) {
+  const result = yield call(ApiSalt.printJob, job.jid);
+  // TODO: error handling?
+  const status = getJobStatusFromPrintJob(result, job.jid);
+  return yield* updateJobStatus(job, status);
```
Comment (carlito-scality) [inline ui/src/ducks/app/salt.js]: why `yield*` instead id `yield call`?

File: ui/src/ducks/app/salt.js [hunk 7]
```diff
@@ -1,31 +1,258 @@
...
+  }
+}
+// }}}
+
+export function* saltSaga() {
+  yield all([
+    fork(manageLocalStorage),
+    fork(garbageCollectJobs),
+    takeEvery(CONNECT_SALT_API, initialize),
+    takeEvery(CONNECT_SALT_API, watchSaltEvents),
```
Comment (carlito-scality) [inline ui/src/ducks/app/salt.js]: I would merge those two in one saga and call `initialize` before `watchSaltEvents ` so we are sure we do not miss active jobs

File: ui/src/ducks/app/salt.js [hunk 8]
```diff
@@ -1,31 +1,258 @@
...
+  status: {},
+  events: [],
+};
+
 export function addJobAction(job) {
-  return { type: ADD_JOB, payload: job };
+  const payload = { ...defaultJob, ...job };
+  if (payload.completedAt !== null) {
+    payload.completed = true;
+  }
```
Comment (carlito-scality) [inline ui/src/ducks/app/salt.js:102]: I wonder if it is necessary here since we will overwrite it in `updateJobStatus`

**Review by carlito-scality — CHANGES_REQUESTED** [review]

File: ui/src/ducks/app/salt.js (hunk 8 above)
Comment (gdemonet) [reply ui/src/ducks/app/salt.js:102]: It's for the `refreshNodes` saga actually, so that they don't show the node as deploying while we wait for an update of the status. We know the job is completed, and the consumer uses `completed`, not `completedAt` :/

File: ui/src/ducks/app/salt.js (hunk 6 above)
Comment (gdemonet) [reply ui/src/ducks/app/salt.js]: :man_shrugging: I read that `yield*` is the idiomatic way of doing this with ES6 generators if you want sequential execution. But true, `call` is useful if we want to test that this saga delegates to another.

File: ui/src/ducks/app/salt.js (hunk 7 above)
Comment (gdemonet) [reply ui/src/ducks/app/salt.js]: Good point! I'll do that

File: ui/src/ducks/app/nodes.js (hunk 4 above)
Comment (gdemonet) [reply ui/src/ducks/app/nodes.js:371]: I actually didn't want to add too much in `localStorage`. But since we don't care too much about "dimensions" when it comes to JSON, I think it actually makes a lot of sense to switch to `type` + some arbitrary number of attributes for consumers to read from, in place of some known format to encode these attributes in a single `name` string.

Good point :)

File: ui/src/ducks/app/salt.js [hunk 9]
```diff
@@ -1,31 +1,267 @@
...
+
+export function setJobStatusAction(jid, status) {
+  return { type: SET_JOB_STATUS, payload: { jid, status } };
+}
+
+// Selectors
+export const allJobsSelector = state => state.app.salt.jobs;
+
+// Sagas
+export function* initialize() {
```
Comment (carlito-scality) [inline ui/src/ducks/app/salt.js]: you need to get a `payload` here and pass it to `yield fork(watchSaltEvents, payload);`

**Review by carlito-scality — CHANGES_REQUESTED** [review]

**Review by carlito-scality — APPROVED** [review]
LGTM
